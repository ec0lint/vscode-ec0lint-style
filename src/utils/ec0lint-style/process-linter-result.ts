import { warningToDiagnostic } from './warning-to-diagnostic';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import { LintDiagnostics, InvalidOptionError } from './types';

/**
 * Processes the results of a ec0lint-style lint run.
 *
 * If ec0lint-style reported any warnings, they are converted to Diagnostics and
 * returned. If the lint results contain raw output in the `output` property, it
 * is also returned.
 *
 * Throws an `InvalidOptionError` for any invalid option warnings reported by
 * ec0lint-style.
 * @param ec0lintStyle The ec0lint-style instance that was used.
 * @param result The results returned by ec0lint-style.
 */
export function processLinterResult(
	ec0lintStyle: Ec0lintStyle.PublicApi,
	{ results, output }: Ec0lintStyle.LinterResult,
): LintDiagnostics {
	if (results.length === 0) {
		return { diagnostics: [] };
	}

	const [{ invalidOptionWarnings, warnings, ignored }] = results;

	if (ignored) {
		return { diagnostics: [] };
	}

	if (invalidOptionWarnings.length !== 0) {
		throw new InvalidOptionError(invalidOptionWarnings);
	}

	const ruleMetadata = new Proxy(
			{},
			{
				get: (_, key: string) => {
					return ec0lintStyle.rules?.[key]?.meta;
				},
			},
		);


	const diagnostics = warnings.map((warning) => warningToDiagnostic(warning, ruleMetadata));

	return output ? { output: output as string, diagnostics } : { diagnostics };
}
