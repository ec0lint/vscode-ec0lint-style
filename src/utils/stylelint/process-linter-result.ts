import { warningToDiagnostic } from './warning-to-diagnostic';
// eslint-disable-next-line node/no-unpublished-import
import type Ec0lintStyle from 'ec0lint-style';
import { LintDiagnostics, InvalidOptionError } from './types';

/**
 * Processes the results of a Stylelint lint run.
 *
 * If Stylelint reported any warnings, they are converted to Diagnostics and
 * returned. If the lint results contain raw output in the `output` property, it
 * is also returned.
 *
 * Throws an `InvalidOptionError` for any invalid option warnings reported by
 * Stylelint.
 * @param stylelint The Stylelint instance that was used.
 * @param result The results returned by Stylelint.
 */
export function processLinterResult(
	stylelint: Ec0lintStyle.PublicApi,
	{ results, output }: Ec0lintStyle.LinterResult,
): LintDiagnostics {
	if (results.length === 0) {
		return { diagnostics: [] };
	}

	const [{ invalidOptionWarnings, warnings, ignored }] = results;

	if (ignored) {
		return { diagnostics: [] };
	}

	if (invalidOptionWarnings.length !== 0) {
		throw new InvalidOptionError(invalidOptionWarnings);
	}

	const ruleMetadata = new Proxy(
			{},
			{
				get: (_, key: string) => {
					return stylelint.rules?.[key]?.meta;
				},
			},
		);


	const diagnostics = warnings.map((warning) => warningToDiagnostic(warning, ruleMetadata));

	return output ? { output: output as string, diagnostics } : { diagnostics };
}
