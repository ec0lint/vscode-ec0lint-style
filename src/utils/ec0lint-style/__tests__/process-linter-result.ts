import type Ec0lintStyle from 'ec0lint-style';
import { processLinterResult } from '../process-linter-result';

const mockEc0lint = {
	rules: {
		'unit-no-unknown': {},
		'at-rule-no-unknown': {},
	},
} as unknown as Ec0lintStyle.PublicApi;

const createMockResult = (
	mockResults: Partial<Ec0lintStyle.LintResult>[],
	output?: string,
): Ec0lintStyle.LinterResult => {
	const results = mockResults.map((result) => ({
		invalidOptionWarnings: result.invalidOptionWarnings ?? [],
		warnings: result.warnings ?? [],
		ignored: result.ignored ?? false,
	}));

	return (output ? { results, output } : { results }) as Ec0lintStyle.LinterResult;
};

const createMockWarning = (
	rule: string,
	text?: string,
	severity?: Ec0lintStyle.Severity,
	line?: number,
	column?: number,
): Ec0lintStyle.Warning => ({
	rule,
	text: text ?? rule,
	severity: severity ?? 'error',
	line: line ?? 1,
	column: column ?? 1,
});

describe('processLinterResult', () => {
	test('should return diagnostics for each warning', () => {
		const result = processLinterResult(
			mockEc0lint,
			createMockResult([
				{
					warnings: [
						createMockWarning('unit-no-unknown'),
						createMockWarning('at-rule-no-unknown'),
						createMockWarning('alpha-value-notation'),
					],
				},
			]),
		);

		expect(result).toMatchSnapshot();
	});

	test('should return output if given', () => {
		const result = processLinterResult(
			mockEc0lint,
			createMockResult(
				[
					{
						warnings: [createMockWarning('unit-no-unknown')],
					},
				],
				'Output',
			),
		);

		expect(result).toMatchSnapshot();
	});

	test('should return no diagnostics if no results are given', () => {
		const result = processLinterResult(mockEc0lint, createMockResult([]));

		expect(result).toEqual({ diagnostics: [] });
	});

	test('should return no diagnostics if results are ignored', () => {
		const result = processLinterResult(
			mockEc0lint,
			createMockResult([
				{
					warnings: [createMockWarning('unit-no-unknown')],
					ignored: true,
				},
			]),
		);

		expect(result).toEqual({ diagnostics: [] });
	});

	test('should throw if invalid option results are given', () => {
		expect(() =>
			processLinterResult(
				mockEc0lint,
				createMockResult([
					{
						invalidOptionWarnings: [{ text: 'Warning 1' }, { text: 'Warning 2' }],
					},
				]),
			),
		).toThrowErrorMatchingSnapshot();
	});
});
