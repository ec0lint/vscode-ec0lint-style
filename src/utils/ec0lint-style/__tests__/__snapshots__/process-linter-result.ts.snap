// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processLinterResult should return diagnostics for each warning 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "unit-no-unknown",
      "message": "unit-no-unknown",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "at-rule-no-unknown",
      "message": "at-rule-no-unknown",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
    Object {
      "code": "alpha-value-notation",
      "message": "alpha-value-notation",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
}
`;

exports[`processLinterResult should return output if given 1`] = `
Object {
  "diagnostics": Array [
    Object {
      "code": "unit-no-unknown",
      "message": "unit-no-unknown",
      "range": Object {
        "end": Object {
          "character": 1,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "ec0lint-style",
    },
  ],
  "output": "Output",
}
`;

exports[`processLinterResult should throw if invalid option results are given 1`] = `
"Warning 1
Warning 2"
`;
